### PR TITLE
Retain exception fields in deserialization

### PR DESCRIPTION
**Pull request by Tim-Brooks**
In ElasticsearchException `type`, `reason`, and `stacktrace` are 
serialized to xcontent. However, when the xcontent is deserialized 
these fields are lost. This PR adds a method that will optionally retain
these fields for usage when the exception is serialized a second time.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@ywelsch @henningandersen This is a POC of how we could retain the information between deserializing a exception from reindexing and deserializing it for http response.

File: server/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -407,7 +423,29 @@ public static ElasticsearchException fromXContent(XContentParser parser) throws
...
+     * return the original exception type (ie NodeClosedException for example) but just wraps
+     * the type, the reason and the cause of the exception. It also recursively parses the
+     * tree structure of the cause, returning it as a tree structure of {@link ElasticsearchException}
+     * instances.
+     *
+     * @param parser the parser
+     * @param retainXContentFields if true, then the exception reason and type will be retained for
+     *                             re-serialization to XContent.
+     */
+    public static ElasticsearchException fromXContent(XContentParser parser, boolean retainXContentFields) throws IOException {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/ElasticsearchException.java]: I think that ideally we should end up not specifying whether to retain or not, it should be default. But we might need this in 7.x if the message needs to stay BWC?

File: server/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 2]
```diff
@@ -1182,4 +1226,18 @@ private static String toUnderscoreCase(String value) {
         return sb.toString();
     }
 
+    private static class XContentRetainingException extends ElasticsearchException {
+
+        private final String fromXContentExceptionType;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/ElasticsearchException.java]: Would be nice to also return this type from the static `getExceptionName` method.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 3]
```diff
@@ -564,6 +564,63 @@ public void testFromXContentWithHeadersAndMetadata() throws IOException {
...
+        ElasticsearchException parsed2;
+        try (XContentParser parser = createParser(builder2)) {
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsed2 = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+
+        assertNotNull(parsed2);
+        assertEquals(parsed2.getMessage(), "Elasticsearch exception [type=exception, reason=foo]");
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: I think we should also validate that builder2's xcontent has `reason: "foo"` and `type: "exception"`.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 4]
```diff
@@ -564,6 +564,63 @@ public void testFromXContentWithHeadersAndMetadata() throws IOException {
...
+
+        cause = (ElasticsearchException) parsed.getCause();
+        assertEquals(cause.getMessage(), "Elasticsearch exception [type=exception, reason=bar]");
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertEquals(cause.getMessage(), "Elasticsearch exception [type=exception, reason=baz]");
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertEquals(cause.getMessage(),
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: Ideally we should also validate that builder2's xcontent has `type: "routing_missing_exception"`

**Review by henningandersen — COMMENTED** [review]
I think this looks like the right direction. I left a few comments to consider/discuss.

**Pull request closed without merging**